src/wallet/service.go

```diff
@@ -84,89 +86,270 @@ func (serv *Service) CreateWallet(wltName string, options Options) (Wallet, erro
+		return err
+	}
+
+	if w.IsEncrypted() {
+		return ErrWalletEncrypted
+	}
+
+	if err := w.lock(password); err != nil {
+		return err
+	}
+
+	// Set the encrypted wallet
+	serv.wallets.set(w)
+
+	return Save(serv.WalletDirectory, w)
+}
+
+// DecryptWallet decrypts wallet with password
+func (serv *Service) DecryptWallet(wltID string, password []byte) error {
+	serv.Lock()
+	defer serv.Unlock()
+	w, err := serv.getWallet(wltID)
+	if err != nil {
+		return err
+	}
+
+	// Returns error if wallet is not encrypted
+	if !w.IsEncrypted() {
+		return ErrWalletNotEncrypted
+	}
+
+	unlockWlt, err := w.unlock(password)
+	if err != nil {
+		return err
+	}
+
+	// Updates the wallet file
+	if err := Save(serv.WalletDirectory, unlockWlt); err != nil {
+		return err
+	}
+
+	// Sets the decrypted wallet in memory
+	serv.wallets.set(unlockWlt)
+	return nil
+}
+
+// EncryptWallets encrypts all wallets.
+// If if any wallet is already encrypted, and this wallet does not decrypt with the password, then fail.
+// otherwise encrypt all unencrypted wallets with this password.
+// at the end of the operation, all wallets should be encrypted with the same password. if they cannot be, then it fails
+func (serv *Service) EncryptWallets(password []byte) error {
+	// Checks if any of wallet already encrypted.
+	serv.Lock()
+	defer serv.Unlock()
+	// Gets wallets that are going to be encrypted
+	var wlts []*Wallet
+	for id, w := range serv.wallets {
+		if w.IsEncrypted() {
+			// Checks if this wallet is encrypted with this password
+			_, err := w.unlock(password)
+			if err != nil {
+				if err == cipher.ErrInvalidPassword {
+					return fmt.Errorf("wallet %s is encrypted with a different password", id)
+				}
+
+				return err
+			}
+
+			continue
+		}
+		wlts = append(wlts, w.clone())
+	}
+
+	// Encrypts the wallets
+	for i := range wlts {
+		if err := wlts[i].lock(password); err != nil {
+			return fmt.Errorf("encrypt wallet %s failed: err: %v", wlts[i].Filename(), err)
+		}
+
+		// Updates the wallet file
+		if err := Save(serv.WalletDirectory, wlts[i]); err != nil {
+			return fmt.Errorf("save wallet %s file failed: %v", wlts[i].Filename(), err)
+		}
+
+		// Updates the wallet in memory
+		serv.wallets.set(wlts[i])
+	}
+
+	return nil
+}
+
+// DecryptWallets decrypts all wallets with password
+// Fails if any encrypted wallet fails to decrypt with the password (In the case
+// of multiple encrypted wallets having different passwords).
+// it does not fail if any wallet is already decrypted
+func (serv *Service) DecryptWallets(password []byte) error {
+	serv.Lock()
+	defer serv.Unlock()
+
+	// Decrypts all wallets
+	var wlts []*Wallet
+	for i := range serv.wallets {
+		w := serv.wallets[i].clone()
+		if !w.IsEncrypted() {
+			continue
+		}
+
+		// Checks if the wallet can be decrypted with the provided password,
+		// fails if can't.
+		decryptedWlt, err := w.unlock(password)
+		if err != nil {
+			return fmt.Errorf("decrypt wallet %s failed: %v", w.Filename(), err)
```

Comment:
log error with this message, but return the original error (same in `Encrypt`), so that the API handlers can check the error and return the correct status code